File: code/91/Java/dp1.java
class Solution {
    public int numDecodings(String s) {
        if (s==null || s.length()==0) return 0;
        int len = s.length();
        int memo[] = new int[len];
        for (int i=0; i<len; i++) memo[i] = -1;
        return count(s.toCharArray(), 0, memo);
    }
    
    private int count(char[] ca, int i, int[] memo) {
        if (i>ca.length) 
            return 0;
        if (i==ca.length) 
            return 1;
        if (memo[i]==-1) {
            int ret;
            if (ca[i]=='0') 
                ret = 0;
            else if (ca[i]=='1' || (ca[i]=='2' && i+1<ca.length && ca[i+1]<='6')) 
                ret = count(ca, i+1, memo)+count(ca, i+2, memo);
            else 
                ret = count(ca, i+1, memo);
            memo[i] = ret;
        }   
        return memo[i];
    }
    
}

File: code/91/Python/dp1.py
class Solution:
    def numDecodings(self, s):
        if not s: return 0
        memo = {}
        def cnt(s, i):
            if i not in memo:
                if i>len(s): ret = 0
                elif i==len(s): ret = 1
                elif s[i]=='0': ret = 0
                elif s[i]=='1' or (s[i]=='2' and i+1<len(s) and s[i+1]<='6'): ret = cnt(s, i+1)+cnt(s,i+2)
                else: ret = cnt(s, i+1) # s[i]>'2'
                memo[i] = ret
            return memo[i]
        return cnt(s, 0)
        